File: app/src/main/java/ru/heatnet/dit/engine/core/Exporter.java
package ru.heatnet.dit.engine.core;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.JsonNodeFactory;
import com.fasterxml.jackson.databind.node.NullNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import org.locationtech.jts.geom.Coordinate;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.Map;

/**
 * Экспорт результата в выходной GeoJSON — строго по §7 техприложения
 * (актуальная редакция).
 *
 * Один файл FeatureCollection (EPSG:4326) содержит объекты ВСЕХ вариантов
 * (variant_id у каждого объекта). Четыре типа объектов (§7.1):
 * heat_network (новые участки), heat_chamber (новые камеры),
 * technical_node, variant_summary (geometry: null).
 * Идентификаторы oks_connection_point и существующих heat_chamber в
 * ссылочных полях и unconnected_oks_ids сохраняют исходный тип (§7.2).
 */
public final class Exporter {

    private Exporter() {
    }

    /** Записать результат: все варианты в один FeatureCollection (EPSG:4326). */
    public static void writeResult(Path path, Model.ContestData data,
                                   List<Pipeline.Variant> variants,
                                   Map<String, Object> summary) {
        int workZone = Crs.zoneOf(data.crsWork);

        ObjectMapper mapper = new ObjectMapper();
        ArrayNode features = mapper.createArrayNode();

        for (Pipeline.Variant variant : variants) {
            String vid = String.valueOf(variant.rank);

            // --- §7.2: новые линейные участки ---
            for (NewSegment seg : variant.newSegments) {
                ObjectNode props = mapper.createObjectNode();
                props.put("id", seg.objectId);
                props.put("object_type", "heat_network");
                props.put("variant_id", vid);
                putNodeId(props, "start_node_id", seg.startNodeId, data);
                putNodeId(props, "end_node_id", seg.endNodeId, data);
                props.put("flow_tph", Math.rint(seg.flowTph * 1000.0) / 1000.0);
                props.put("diameter", (int) seg.diameterMm);
                props.put("length", Math.rint(seg.lengthM * 10.0) / 10.0);
                props.put("laying_method", seg.method);   // base | special
                if (seg.depthStart != null) {
                    props.put("depth_start", seg.depthStart);
                } else {
                    props.set("depth_start", NullNode.getInstance());
                }
                if (seg.depthEnd != null) {
                    props.put("depth_end", seg.depthEnd);
                } else {
                    props.set("depth_end", NullNode.getInstance());
                }
                props.put("cost", Math.rint(seg.costRub * 100.0) / 100.0);
                features.add(feature(mapper, lineTo4326(seg.coords, workZone), props));
            }

            // --- §7.2: новые камеры ---
            for (Pipeline.NewChamber ch : variant.newChambers) {
                ObjectNode props = mapper.createObjectNode();
                props.put("id", ch.objectId);
                props.put("object_type", "heat_chamber");
                props.put("variant_id", vid);
                props.put("diameter", (int) ch.diameterMm);
                props.put("cost", Math.rint(ch.costRub * 100.0) / 100.0);
                features.add(feature(mapper, pointTo4326(ch.point, workZone), props));
            }

            // --- §7.2: технические узлы ---
            for (TechnicalNode node : variant.techNodes) {
                ObjectNode props = mapper.createObjectNode();
                props.put("id", node.objectId);
                props.put("object_type", "technical_node");
                props.put("variant_id", vid);
                props.put("reason", node.reason); // доп. поле (§7: допускается)
                features.add(feature(mapper, pointTo4326(node.point, workZone), props));
            }

            // --- §7.2: сводка по варианту (geometry: null) ---
            ObjectNode props = mapper.createObjectNode();
            props.put("id", "summary-" + vid);
            props.put("object_type", "variant_summary");
            props.put("variant_id", vid);
            props.put("rank", variant.rank);
            props.put("construction_cost", variant.costs.get("construction_cost"));
            props.put("chamber_construction_cost",
                    variant.costs.get("chamber_construction_cost"));
            props.put("existing_chamber_tie_in_count",
                    variant.costs.get("existing_chamber_tie_in_count").intValue());
            props.put("existing_chamber_tie_in_cost",
                    variant.costs.get("existing_chamber_tie_in_cost"));
            props.put("unconnected_penalty", variant.costs.get("unconnected_penalty"));
            props.put("calculated_cost", variant.costs.get("calculated_cost"));
            props.put("new_network_length", variant.lengths.get("new_network_length"));
            props.put("score", variant.score);
            ArrayNode unc = mapper.createArrayNode();
            for (Pipeline.Unconnected u : variant.unconnected) {
                // §7.2: тип идентификатора сохраняется как во входных данных
                Model.ConnectionTarget t = data.targets.get(u.buildingId);
                if (t != null && t.idRaw != null && (t.idRaw.isNumber() || t.idRaw.isTextual())) {
                    unc.add(t.idRaw);
                } else {
                    unc.add(u.buildingId);
                }
            }
            props.set("unconnected_oks_ids", unc);
            features.add(feature(mapper, null, props));
        }

        ObjectNode root = mapper.createObjectNode();
        root.put("type", "FeatureCollection");
        root.set("features", features);

        try {
            Path parent = path.getParent();
            if (parent != null) {
                Files.createDirectories(parent);
            }
            mapper.writeValue(path.toFile(), root);
        } catch (IOException e) {
            throw new IllegalStateException("не удалось записать результат: " + e.getMessage(), e);
        }
    }

    // ------------------------------------------------------------------

    /**
     * Ссылка на узел (§7.2): для oks_connection_point и существующих
     * heat_chamber — исходный id с сохранением типа; для новых узлов —
     * внутренний строковый идентификатор.
     */
    private static void putNodeId(ObjectNode props, String key, String nodeId,
                                  Model.ContestData data) {
        if (nodeId == null) {
            props.set(key, NullNode.getInstance());
            return;
        }
        Model.ConnectionTarget t = data.targets.get(nodeId);
        if (t != null && t.idRaw != null && (t.idRaw.isNumber() || t.idRaw.isTextual())) {
            props.set(key, t.idRaw);
            return;
        }
        Model.Chamber ch = data.chambers.get(nodeId);
        if (ch != null && ch.idRaw != null && (ch.idRaw.isNumber() || ch.idRaw.isTextual())) {
            props.set(key, ch.idRaw);
            return;
        }
        props.put(key, nodeId);
    }

    private static ObjectNode feature(ObjectMapper mapper, JsonNode geom, ObjectNode props) {
        ObjectNode f = mapper.createObjectNode();
        f.put("type", "Feature");
        if (geom != null) {
            f.set("geometry", geom);
        } else {
            f.set("geometry", NullNode.getInstance());
        }
        f.set("properties", props);
        return f;
    }

    // ------------------------------------------------------------------
    // GeoJSON mapping (EPSG:4326)

    private static JsonNode lineTo4326(List<Coordinate> coords, int zone) {
        ArrayNode arr = JsonNodeFactory.instance.arrayNode();
        for (Coordinate c : coords) {
            arr.add(posTo4326(c.x, c.y, zone));
        }
        ObjectNode g = JsonNodeFactory.instance.objectNode();
        g.put("type", "LineString");
        g.set("coordinates", arr);
        return g;
    }

    private static JsonNode pointTo4326(org.locationtech.jts.geom.Point p, int zone) {
        ObjectNode g = JsonNodeFactory.instance.objectNode();
        g.put("type", "Point");
        g.set("coordinates", posTo4326(p.getX(), p.getY(), zone));
        return g;
    }

    private static ArrayNode posTo4326(double x, double y, int zone) {
        double[] lonlat = Crs.inverse(x, y, zone);
        ArrayNode pos = JsonNodeFactory.instance.arrayNode();
        pos.add(lonlat[0]);
        pos.add(lonlat[1]);
        return pos;
    }
}

File: oracle/app/engine/exporter.py
"""Экспорт результата в выходной GeoJSON — строго по разделу 10.

Один файл FeatureCollection (EPSG:4326) содержит объекты ВСЕХ вариантов
(variant_id у каждого объекта). Ровно семь типов объектов, у каждого —
только свой набор атрибутов (таблицы 10.1–10.7), без посторонних полей
и без metadata:

  heat_network                 — новый линейный участок;
  tie_in                       — точка врезки в существующую сеть;
  heat_network_reconstruction  — реконструируемая часть существующего участка;
  heat_chamber                 — новая тепловая камера;
  heat_chamber_reconstruction  — реконструкция существующей камеры;
  technical_node               — технический узел;
  variant_summary              — сводная запись варианта (geometry: null).
"""

from __future__ import annotations

import json
from pathlib import Path

from pyproj import Transformer
from shapely.geometry import LineString, mapping
from shapely.ops import transform as shp_transform

from .reconstruction import Reconstruction


def write_result(path: Path, data, variants: list, summary: dict) -> None:
    back = Transformer.from_crs(data.crs_work, "EPSG:4326", always_xy=True).transform

    def to4326(geom):
        return shp_transform(back, geom)

    def feature(geom, props):
        return {
            "type": "Feature",
            "geometry": mapping(to4326(geom)) if geom is not None else None,
            "properties": props,
        }

    def chamber_existing_dn(ch) -> float:
        if ch.diameter_mm:
            return ch.diameter_mm
        adj = [s.diameter_mm or 0.0 for s in data.segments.values()
               if s.next_object_id == ch.object_id]
        return max(adj) if adj else 0.0

    features = []

    for variant in variants:
        vid = str(variant.get("rank", 0))

        # --- §10.1: новые линейные участки ---
        for seg in variant["new_segments"]:
            features.append(feature(LineString(seg.coords), {
                "id": seg.object_id,
                "object_type": "heat_network",
                "variant_id": vid,
                "start_node_id": seg.start_node_id,
                "end_node_id": seg.end_node_id,
                "flow_tph": round(seg.flow_tph, 3),
                "diameter": int(seg.diameter_mm),
                "length": round(seg.length_m, 1),
                "laying_method": seg.method,   # base | special
                "depth_start": seg.depth_start,  # null в 2D-задаче (§8.1)
                "depth_end": seg.depth_end,
                "cost": round(seg.cost_rub, 2),
            }))

        # --- §10.2: точки врезки ---
        for tap in variant["taps"]:
            if tap.kind == "existing_chamber":
                ch = data.chambers[tap.chamber_id]
                existing_id = tap.chamber_id
                existing_type = "heat_chamber"
                existing_dn = chamber_existing_dn(ch)
            else:
                seg0 = data.segments[tap.segment_id]
                existing_id = tap.segment_id
                existing_type = "heat_network"
                existing_dn = seg0.diameter_mm or 0.0
            features.append(feature(tap.point, {
                "id": tap.node_id,
                "object_type": "tie_in",
                "variant_id": vid,
                "existing_object_id": existing_id,
                "existing_object_type": existing_type,
                "existing_diameter": int(existing_dn),
                "required_diameter": int(tap.required_dn),
                "cost": round(tap.tap_cost_rub, 2),
            }))

        # --- §10.3: реконструкция линейных участков ---
        recon: Reconstruction = variant["recon"]
        for rs in recon.segments:
            features.append(feature(rs.geom, {
                "id": rs.object_id,
                "object_type": "heat_network_reconstruction",
                "variant_id": vid,
                "existing_object_id": rs.existing_object_id,
                "existing_flow_tph": rs.existing_flow,
                "added_flow_tph": rs.added_flow,
                "calculated_flow_tph": rs.calculated_flow,
                "existing_diameter": int(rs.existing_dn),
                "required_diameter": int(rs.required_dn),
                "length": rs.length_m,
                "cost": rs.cost_rub,
            }))

        # --- §10.4: новые камеры ---
        for ch in variant["new_chambers"]:
            features.append(feature(ch["point"], {
                "id": ch["object_id"],
                "object_type": "heat_chamber",
                "variant_id": vid,
                "diameter": int(ch["diameter_mm"]),
                "cost": round(ch["cost_rub"], 2),
            }))

        # --- §10.5: реконструкция существующих камер ---
        for rc in recon.chambers:
            features.append(feature(data.chambers[rc.existing_object_id].geom, {
                "id": rc.object_id,
                "object_type": "heat_chamber_reconstruction",
                "variant_id": vid,
                "existing_object_id": rc.existing_object_id,
                "existing_diameter": int(rc.existing_dn),
                "required_diameter": int(rc.required_dn),
                "cost": rc.cost_rub,
            }))

        # --- §10.6: технические узлы ---
        for node in variant["tech_nodes"]:
            features.append(feature(node.point, {
                "id": node.object_id,
                "object_type": "technical_node",
                "variant_id": vid,
            }))

        # --- §10.7: сводная запись варианта (geometry: null) ---
        costs = variant["costs"]
        lengths = variant["lengths"]
        features.append(feature(None, {
            "id": f"summary-{vid}",
            "object_type": "variant_summary",
            "variant_id": vid,
            "rank": int(variant.get("rank", 0)),
            "construction_cost": costs["construction_cost"],
            "chamber_construction_cost": costs["chamber_construction_cost"],
            "tie_in_cost": costs["tie_in_cost"],
            "reconstruction_cost": costs["reconstruction_cost"],
            "chamber_reconstruction_cost": costs["chamber_reconstruction_cost"],
            "unconnected_penalty": costs["unconnected_penalty"],
            "calculated_cost": costs["calculated_cost"],
            "new_network_length": lengths["new_network_length"],
            "reconstruction_length": lengths["reconstruction_length"],
            "length": lengths["length"],
            "score": variant.get("score"),
            "unconnected_oks_ids": list(variant["unconnected_ids"]),
        }))

    collection = {
        "type": "FeatureCollection",
        "features": features,
    }

    path = Path(path)
    path.parent.mkdir(parents=True, exist_ok=True)
    with path.open("w", encoding="utf-8") as fh:
        json.dump(collection, fh, ensure_ascii=False)
